### PR TITLE
BaSP-TG-06: projects 

### PR DESCRIPTION
**Pull request by matescarabino**
- Main structure
- getAll, getById and create method working
- postman collection

### CODE REVIEW

File: src/controllers/projects.js [hunk 1]
```diff
@@ -1,189 +1,71 @@
...
-const getAllProjects = (req, res) => {
-  res.status(200).json(projects);
-};
-
-// getActive projects
-const getActiveProjects = (req, res) => {
-  const foundProjects = projects.filter((project) => project.isDeleted === false);
-  res.status(200).json(foundProjects);
-};
+// GetAll
```
Comment (lmalvasia) [inline src/controllers/projects.js:3]: Remove this comment

File: src/controllers/projects.js [hunk 2]
```diff
@@ -1,189 +1,71 @@
...
+const getAll = async (req, res) => {
+  try {
+    const projects = await Projects.find();
```
Comment (lmalvasia) [inline src/controllers/projects.js:6]: filter the data based on the query params:
```suggestion
    const projects = await Projects.find(req.query);
```

File: src/controllers/projects.js [hunk 3]
```diff
@@ -1,189 +1,71 @@
...
-  let projectArray = projects;
-  const queriesArray = Object.keys(req.query);
-  queriesArray.forEach((query) => {
-    if (query !== 'id' && query !== 'name' && query !== 'startDate' && query !== 'endDate' && query !== 'clientName' && query !== 'employees') {
-      res.status(400).json({
-        Error: 'Applied filter does not exist',
-      });
-    }
-  });
+// GetById
```
Comment (lmalvasia) [inline src/controllers/projects.js:21]: Remove this comment

File: src/controllers/projects.js [hunk 4]
```diff
@@ -1,189 +1,71 @@
...
+const getById = async (req, res) => {
```
Comment (lmalvasia) [inline src/controllers/projects.js:22]:
```suggestion
const getProjectById = async (req, res) => {
```

File: src/controllers/projects.js [hunk 5]
```diff
@@ -1,189 +1,71 @@
...
-  res.status(200).json(projects);
-};
-
-// getActive projects
-const getActiveProjects = (req, res) => {
-  const foundProjects = projects.filter((project) => project.isDeleted === false);
-  res.status(200).json(foundProjects);
-};
+// GetAll
+const getAll = async (req, res) => {
```
Comment (lmalvasia) [inline src/controllers/projects.js:4]:
```suggestion
const getAdmins = async (req, res) => {
```

File: src/controllers/projects.js [hunk 6]
```diff
@@ -1,189 +1,71 @@
...
-      res.status(400).json({
-        Error: 'Applied filter does not exist',
-      });
-    }
-  });
+// GetById
+const getById = async (req, res) => {
+  try {
+    const { id } = req.params;
+    const projects = await Projects.findById(id);
```
Comment (lmalvasia) [inline src/controllers/projects.js:25]:
```suggestion
    const project = await Projects.findById(id);
```
The findById method could return a null/undefined value. So you have to add an extra validation and check the value of the project variable.
If it's undefined you have to return a different response.

File: src/controllers/projects.js [hunk 7]
```diff
@@ -1,189 +1,71 @@
...
-          emptyBody,
-          employeeDoesNotExists: !employeeExists,
-          employeeRate: 'Employee Rate cannot be negative',
-        },
-        error: true,
-      });
-    }
-  } else {
-    res.status(404).json({ msg: `Project not found by id: ${projectId}`, error: true });
+// Post
```
Comment (lmalvasia) [inline src/controllers/projects.js:40]: Remove this comment

File: src/controllers/projects.js [hunk 8]
```diff
@@ -1,189 +1,71 @@
...
 };
 
-module.exports = {
-  getAllProjects,
-  getActiveProjects,
-  getProjectById,
-  updateProjects,
-  deleteProjects,
-  assignEmployee,
+// Exports
```
Comment (lmalvasia) [inline src/controllers/projects.js:66]: Remove this comment

File: src/controllers/projects.js [hunk 9]
```diff
@@ -1,189 +1,71 @@
...
+export default {
```
Comment (lmalvasia) [inline src/controllers/projects.js:67]: Don't use export default. You have to export each function individually.

File: src/validations/projects.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import Joi from 'joi';
+
+const validateCreation = (req, res, next) => {
```
Comment (lmalvasia) [inline src/validations/projects.js:3]:
```suggestion
const validateProjectBody = (req, res, next) => {
```

File: src/validations/projects.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    return res.status(400).json({
+      message: `There was an error: ${validation.error.details[0].message}`,
+      data: undefined,
+      error: true,
+    });
+  }
+  return next();
+};
+
+export default {
```
Comment (lmalvasia) [inline src/validations/projects.js:31]: export default validateProjectBody;

Comment (lmalvasia) [reply src/validations/projects.js:31]: Fix this one

**Review by lmalvasia — DISMISSED** [review]

**Review by kolodani — APPROVED** [review]
great job

**Pull request merged**